conmon: implement logging

cyphar opened the pull request:
This adds a very simple implementation of logging within conmon, where
every buffer read from the masterfd of the container is also written to
the log file (with errors during writing to the log file ignored). It also
includes changes to how we compute the logPath for a container.

The WIP part of this PR is that conmon currently doesn't really handle
containers where `terminal: false`. This is going to be not-very-nice to
fix because the current runC API is broken (which is why we need to
make sure that the current state of opencontainers/runtime-spec#513
is **fixed**). However, while that's being sorted (and I work on the
relevant console changes in runC) we can solve it with a `fork+execve`
though no doubt there will be some fun issues with the exit code.

Closes #24.

Signed-off-by: Aleksa Sarai asarai@suse.de

cyphar commented:
Yeah, the CLA still isn't signed but I though I'd open a PR so nobody else re-does work I've already done. It also opens the door for testing.

mrunalp commented:
Sweet! I'll test this out.

cyphar commented:
I've just pushed a version where `terminal=false` logging actually works. It's all a little bit dodgy at the moment, but now (for the first time) we actually have full logging to a file. :D

mrunalp commented:
@cyphar Fails tests. I will try manual testing next. 
```
1..10
not ok 1 ctr remove
# (in test file ./ctr.bats, line 35)
#   `[ "$status" -eq 0 ]' failed
# time="2016-10-19T10:54:45-07:00" level=info msg="Starting reaper" 
# E1019 10:54:45.599650    7480 ocicni.go:136] error updating cni config: No networks found in /etc/cni/net.d
# time="2016-10-19T10:54:45-07:00" level=debug msg="sandboxes: map[]" 
# time="2016-10-19T10:54:45-07:00" level=debug msg="containers: &{map[] {{0 0} 0 0 0 0}}" 
# time="2016-10-19T10:54:45-07:00" level=info msg="Signal received: child exited" 
# time="2016-10-19T10:54:45-07:00" level=info msg="Signal received: child exited" 
# time="2016-10-19T10:54:45-07:00" level=debug msg="setting sandbox infra container's log_path" ctr.logfile=d35e7bc3a1bd8111f6c904dd72a5c65c6ffbde65ab170f08118391b2a7d8c46d log_path="/var/log/ocid/pods/5322d4b3450a79798e1a34f9c676fc2005bac828d09fbf7ba382
1b5cac6fc075/d35e7bc3a1bd8111f6c904dd72a5c65c6ffbde65ab170f08118391b2a7d8c46d.log" sbox.logdir="/var/log/ocid/pods/5322d4b3450a79798e1a34f9c676fc2005bac828d09fbf7ba3821b5cac6fc075" 
# time="2016-10-19T10:54:45-07:00" level=debug msg="copying infra rootfs binary: /root/gosrc/src/github.com/kubernetes-incubator/cri-o/pause/pause -> /tmp/tmp.3Qj9yykBiS/ocid/graph/vfs/pause/rootfs/pause" 
# time="2016-10-19T10:54:45-07:00" level=debug msg="running conmon: /root/gosrc/src/github.com/kubernetes-incubator/cri-o/conmon/conmon" args=[-c default-podsandbox1-0-infra -r /usr/local/sbin/runc -l /var/log/ocid/pods/5322d4b3450a79798e1a34f9c676fc2005
bac828d09fbf7ba3821b5cac6fc075/d35e7bc3a1bd8111f6c904dd72a5c65c6ffbde65ab170f08118391b2a7d8c46d.log] 
# [conmon:i]: container PID: 7612
# [conmon:e] Failed to add stdin to epoll Operation not permitted
```

mrunalp commented:
pod create/ctr create/ctr start work fine. However I don't see any logs in the container.log file :/
Also ctr stop hangs..

cyphar commented:
@mrunalp How did you run it? You need to give an absolute path to the log directory in your configs -- otherwise odd things happen to the cwd that I haven't figured out yet. For example, the test data is currently broken (though I think I included a fix) because it sets the `log_directory` to `"."`.

mrunalp commented:
@cyphar I'll try with setting the log_directory. Thanks!

cyphar commented:
@mrunalp Note: if you set `log_directory: ""` then my code will automatically put all of the logs inside `/var/log/ocid/pod/<pod id>/<ctr id>.log`. The kubelet will always tell us a log path, but I thought this might be useful just for our own testing.

mrunalp commented:
@cyphar I set pod log_directory to `/var/log/podsandbox1` and container log_path to `container.log` The files do get created but I don't see any logs in `/var/log/podsandbox1/container.log`.

cyphar commented:
Yeah, that's probably a bug with how I'm handling things. I'll look at this later today.

cyphar commented:
@mrunalp Just to make sure, you did this right (I actually made this mistake a few times):

```
% sbox_id=$(sudo ./ocic pod create --config test/testdata/sandbox_config.json --name sandbox1)
% ctr_id=$(sudo ./ocic ctr create --config test/testdata/container_config.json --pod 725b26dbf73ad153ca6ac7a6bc69db2872bcb98e495c7836fa34c4bc62219445)
% sudo ./ocic ctr start --id $ctr_id
```

After making the changes you mentioned. It works for me:

```
% sudo cat /var/log/ocid/podsandbox1/container.log
bin
boot
data
dev
etc
home
lib
lib64
media
mnt
opt
proc
root
run
sbin
srv
sys
tmp
usr
var
```

mrunalp commented:
@cyphar I have been testing container_redis.json. I'll try with the other config and check whether redis is writing to stdout or not.

mrunalp commented:
@cyphar I just got around to testing this again. Your example with ls works fine but redis-server doesn't work with or without tty.

cyphar commented:
@mrunalp I've got a [branch](https://github.com/cyphar/cri-o/tree/conmon-logging-consolesocket) with the `--console-socket` implementation, and found some bugs in the runC `--console-socket` stuff (nothing major). Currently this branch works with `conmon` if you run it by itself, but for some reason it doesn't work under `ocid`.

mrunalp commented:
Okay, just retested this. redis works without tty but no logs with tty.

runcom commented:
Ping @cyphar any update?

mrunalp commented:
We are close to getting console rewrite merged in runc which worked last I tested so can just wait for that.



> On Nov 24, 2016, at 10:31 AM, Antonio Murdaca <notifications@github.com> wrote:
> 
> Ping @cyphar any update?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

runcom commented:
@mrunalp thanks, wasn't aware of that

feiskyer commented:
What's the status of this?

cyphar commented:
The same as above. There is a console rewrite PR in runC pending to be merged (opencontainers/runc#1018) -- when you apply that patch to runC this PR works properly. Otherwise there are issues with `tty: false` containers and I don't really want to debug old code that is going to be replaced soon.

cyphar commented:
opencontainers/runc#1018 has been merged. I will rebase this soon.

runcom commented:
@cyphar any update on this one :angel:

cyphar commented:
@runcom Sorry, `umoci` got in the way. I will work on this today / tomorrow. Promise. If not, I give you permission to spam me on twitter.

cyphar commented:
Hmmm. I need to figure out where all of my `logDir` code changes need to be moved to now.

cyphar commented:
Local integration still is broken, with `Is a directory` errors. I'm going to have to figure out what change I've dropped.

runcom commented on conmon/cmsg.c:
```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2016 SUSE LLC
```
do we keep this?

cyphar replied:
The code comes from runC. If you want I can drop it (though personally I prefer having license headers in all source files).

runcom replied:
@mrunalp wdyt?

jawnsy replied:
This may need to be referred to legal. If the code was written during work hours or on equipment provided by SUSE LLC, they will generally own copyright (and this is usually stipulated as part of employment agreements). It's not permissible to change the statement without approval from the copyright owner (being the legal representative acting on behalf of SUSE LLC). This is especially true if the code was extracted from runC and already carries this sort of statement. This is generally why many projects require a Contributor License Agreement or at least a Developer Certificate of Origin.

@cyphar I don't have any personal opinion on individual copyright statements in files, but I will say that some vendors (Debian in particular) require a comprehensive machine-readable copyright file, which includes the names of all rights holders, and having additional copyright statements means more work for packagers. Just something to consider.

cyphar replied:
@jawnsy In general, it's just the style within SUSE (and the GNU project, and Linux, and systemd, and ...) to include a copyright statement in all files (and also RedHat AFAIK, though in the Go community people appear to not like including copyright statements in comments because comments actually affect compilation). That's why I included it in the original file, so sorry for any inconvenience.

jawnsy replied:
@cyphar I'm not arguing with common practice, but someone (a Debian contributor, maintainer, or developer) needs to collate all those copyright statements, so it's manual work for somebody; for example: http://metadata.ftp-master.debian.org/changelogs/main/s/systemd/systemd_232-8_copyright

cyphar replied:
Well, you could actually automate it because according to US copyright laws only strings formatted in a certain way are considered to be a legally enforceable copyright statements.

However, to make this simpler do you want to do something like how Kubernetes handles it?

```
Copyright 2015 The Kubernetes Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

jawnsy replied:
As these projects (SUSE, Debian, Fedora) are not US-only, we do have to worry about international copyright law as well, and related concepts like [moral rights](https://en.wikipedia.org/wiki/Moral_rights). 

I think the Kubernetes statement is good, and would definitely prefer something like that. A similar, generic, all-encompassing statement is used for the Gradle project, and this seems to be accepted by the Debian folk: http://metadata.ftp-master.debian.org/changelogs/main/g/gradle/gradle_3.2.1-1_copyright

runcom commented on server/server.go:
```diff
@@ -194,7 +194,7 @@ func (s *Server) loadSandbox(id string) error {
 		return err
 	}
-	scontainer, err := oci.NewContainer(m.Annotations["ocid/container_id"], cname, sandboxPath, sandboxPath, sb.netNs(), labels, annotations, nil, nil, id, false)
+	scontainer, err := oci.NewContainer(m.Annotations["ocid/container_id"], cname, sandboxPath, m.Annotations["ocid/log_path"], sb.netNs(), labels, annotations, nil, nil, id, false)
```
I feel we can remove `m.Annotations["..."]` and just use `sb.annotations` inside `oci.NewContainer`

runcom commented:
`conmon/cmsg.o` should be added to `.gitignore` I guess

cyphar commented:
@runcom No, I think it's related to the `EISDIR` errors I got before -- it's because the log directories given by the configuration from k8s (or our tests) are not the values you might expect. I'll re-read the relevant interface docs to jog my memory.

cyphar commented:
`EISDIR` is gone. Unfortunately, I'm having issues running the test suite on my local machine -- things appear to be broken on openSUSE.

cyphar commented:
@mrunalp @runcom I've tested it manually, and logging now works. I'm having trouble running the test suite on openSUSE still, but the logging code definitely appears to work.

mrunalp commented:
@cyphar Nice! I will test this out.

mrunalp commented:
Creating sandbox fails with this PR:
```
[conmon:i]: about to waitpid: 13390
[conmon:e]: Failed to create container: exit status 1
2017/01/05 16:13:43 grpc: Server failed to encode response proto: Marshal called with nil
```

mrunalp commented:
I tested this again. So, it works for pre-console-socket runc with tty set to false. Sandbox creation fails when I switch to runc master.

runcom commented on server/sandbox_run.go:
```diff
@@ -151,6 +145,28 @@ func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
+		var wd string
+		logrus.Warnf("requested logDir for sbox id %s is a relative path: %s", id, logDir)
+
+		// XXX: What is a sane thing to do here if kubernetes really intended
```
might be sane to ask this upstream in kubernetes, maybe it shouldn't send us relative paths, in that case we could error out.

feiskyer replied:
This wouldn't happen, see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/helpers.go#L222 and 
`podLogsRootDirectory = "/var/log/pods"`

jawnsy replied:
If it's something that "should never happen", I'm +1 to @runcom's comment: we should error out

mrunalp commented:
We need to get this one in next. Needs rebase.

cyphar commented:
@mrunalp I will investigate this because it used to work with post-console-PR runC. So something has broken since then...

mrunalp commented:
Any updates?

cyphar commented:
@mrunalp I've rebased again, but I'm really confused why it isn't working for you -- it works for me if I manually run some logging tests (though as I've mentioned before we're still having issues with running the test suite on openSUSE).

runcom commented:
@cyphar any way this can be tested with bats? I'm going to test this out on fedora

cyphar commented:
@runcom Yes, I've modified the test data so that it should work properly with `bats` "out of the box" -- though I will have to add some integration tests to make sure that the log file contains the right output.

cyphar commented:
Ah, tests are failing because the `ExecSync` changes to `conmon` (which need to be rewritten when we switch to a streaming server btw) don't provide a `--log-path`...

mrunalp commented:
I'll test this out again. Thanks!

mrunalp commented:
It works for me! :) Now we need to fix the test issues that @cyphar pointed out.

runcom commented:
works for me as well :+1:

cyphar commented:
Alright. Let me work on fixing the `ExecSync` and adding tests for logging.

I also just found out that the `btrfs` storage driver has some bugs upstream (read: Docker), so I think that the issues I've been having with testing are a result of those issues -- so I'll start testing with devicemapper while we try to fix `btrfs`.

mrunalp commented:
One more thing: we need to decorate the log entries. For details see https://github.com/kubernetes/community/blob/master/contributors/design-proposals/kubelet-cri-logging.md

cyphar commented:
I'm quite nervous about implementing log decoration in C. Also, something that we should consider is that currently `--console-socket` sets both `stdout` and `stderr` to the same file descriptor (this was also true for the previous terminal implementation). How should we handle that?

mrunalp commented:
> Sure @mrunalp, though I'm quite nervous about implementing log decoration in C
Yeah, agree that it will be tricky.
>  Also, something that we should consider is that currently --console-socket sets both stdout and stderr to the same file descriptor (this was also true for the previous terminal implementation). How should we handle that?
Yeah, with terminal equal to true, there is no simple way to distinguish between stdout/stderr that I can think of.  We would have to implement something like tee. (The other k8s runtimes must have the same issue.)

For terminal false, we can dup stdout and stderr to different file fds or pipes.

cyphar commented:
@mrunalp 

> Yeah, agree that it will be tricky

According to the doc you linked, we can do this change incrementally. First get this code merged and then do decoration in a separate patch?

> For terminal false, we can dup stdout and stderr to different file fds or pipes.

That is something I'll have to implement. But until we add decoration it's not really necessary at the moment.

mrunalp commented:
Yeah, sounds good. We can add decoration in follow on PRs.


> On Feb 25, 2017, at 9:31 PM, Aleksa Sarai <notifications@github.com> wrote:
> 
> @mrunalp
> 
> Yeah, agree that it will be tricky
> 
> According to the doc you linked, we can do this change incrementally. First get this code merged and then do decoration in a separate patch?
> 
> For terminal false, we can dup stdout and stderr to different file fds or pipes.
> 
> That is something I'll have to implement. But until we add decoration it's not really necessary at the moment.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mrunalp commented:
@cyphar Any updates?

cyphar commented:
@mrunalp The `ExecSync` tests pass now, and I've added tests for logging. This required updating runC (we can't use `v1.0.0-rc3` because the new runC now supports `-rc5` of the runtime spec and we don't).

mrunalp commented:
@cyphar Awesome. I'll work on updating our dependencies to rc5 of the spec. Meanwhile we can review and get this in.

cyphar commented:
All of the tests pass, the only error is from `gocyclo` -- which really shouldn't be a fatal error IMO because of how often stuff like this breaks CI.

mrunalp commented:
Works fine but the execsync test just hangs :/

cyphar commented:
@mrunalp The integration test doesn't hang on Travis. What version of runC are you using? Maybe if you apply the changes Michael made to `--console-socket` it will work better?

mrunalp commented:
@cyphar `make localintegration` was hanging for me but it was with a version before Micheal's changes. I will try it with those changes included.

mrunalp commented:
@cyphar okay, the tests are working for me! :) We can increase the cyclo line count or break it up into some functions. 
@sameo @runcom Could you guys test/review this as well?

mrunalp commented:
I tested with runc version `b7932a2e07b04d956f9e64d2d01d39fee651c292`

runcom commented:
> @cyphar okay, the tests are working for me! :) We can increase the cyclo line count or break it up into some functions.

testing this out now, I'm ok increasing the cyclo and/or breaking up the function

runcom commented:
LGTM (pending gocyclo failure)

runcom commented:
Testing this with k8s has issues though, around starting pods.

- CRI-O debug logs https://runcom.red/debug.log
- kubelet logs https://runcom.red/kubelet.log
- runc does list the container which supposedly is giving that not found error (I don't know why)
- testing with @mrunalp's runc commit above
- testing with k8s master (from today)

runcom commented:
more logs:
```
[conmon:i]: about to waitpid: 9640
[conmon:i]: container PID: 9648
[conmon:i]: about to waitpid: 11382
[conmon:e]: Failed to create container: exit status 1
```
```
DEBU[2017-03-27 17:47:38.934771565+02:00] ListContainersRequest &ListContainersRequest{Filter:&ContainerFilter{Id:,State:&ContainerStateValue{State:CONTAINER_RUNNING,},PodSandboxId:,LabelSelector:map[string]string{},},}
DEBU[2017-03-27 17:47:38.937151281+02:00] updating sandbox and container information
WARN[2017-03-27 17:47:38.948936325+02:00] could not load new sandbox container 0eaf63282a3bc93f8ed0db66d5355b1274230441572b5c3b3669404a3e845608: error updating status for container 0eaf63282a3bc93f8ed0db66d5355b1274230441572b5c3b3669404a3e845608: error getting container state for default-nginx-2371676037-fnrc6-0-nginx-0: exit status 1: "container \"default-nginx-2371676037-fnrc6-0-nginx-0\" does not exist\n", ignoring
DEBU[2017-03-27 17:47:38.951284853+02:00] ListContainersResponse: &ListContainersResponse{Containers:[],}
DEBU[2017-03-27 17:47:39.664683206+02:00] ListPodSandboxRequest &ListPodSandboxRequest{Filter:nil,}
DEBU[2017-03-27 17:47:39.666988118+02:00] updating sandbox and container information
WARN[2017-03-27 17:47:39.675824110+02:00] could not load new sandbox container 0eaf63282a3bc93f8ed0db66d5355b1274230441572b5c3b3669404a3e845608: error updating status for container 0eaf63282a3bc93f8ed0db66d5355b1274230441572b5c3b3669404a3e845608: error getting container state for default-nginx-2371676037-fnrc6-0-nginx-0: exit status 1: "container \"default-nginx-2371676037-fnrc6-0-nginx-0\" does not exist\n", ignoring
```

mrunalp commented:
This is what I see in the logs:
```
[root@localhost cri-o]# cat /var/log/pods/cbe8a75a-1317-11e7-a80e-54ee75105a2b/nginx_5.log
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (6: No such device or address)
2017/03/27 18:07:14 [emerg] 1#1: open() "/var/log/nginx/error.log" failed (6: No such device or address)
```

mrunalp commented:
okay, here is more info:
```
[root@localhost userdata]# ls -l /var/lib/ocid/devicemapper/mnt/bda94d7cc06b6f23180c1b8e9c69e19e49aa6d37098a78d4f9fd05fd1a91410e/rootfs/var/log/nginx/error.log 
lrwxrwxrwx. 1 root root 11 Feb 28 07:15 /var/lib/ocid/devicemapper/mnt/bda94d7cc06b6f23180c1b8e9c69e19e49aa6d37098a78d4f9fd05fd1a91410e/rootfs/var/log/nginx/error.log -> /dev/stderr
[root@localhost userdata]# ls -l /var/lib/ocid/devicemapper/mnt/bda94d7cc06b6f23180c1b8e9c69e19e49aa6d37098a78d4f9fd05fd1a91410e/rootfs/var/log/nginx/access.log 
lrwxrwxrwx. 1 root root 11 Feb 28 07:15 /var/lib/ocid/devicemapper/mnt/bda94d7cc06b6f23180c1b8e9c69e19e49aa6d37098a78d4f9fd05fd1a91410e/rootfs/var/log/nginx/access.log -> /dev/stdout
```

which is setup here:
```
https://github.com/nginxinc/docker-nginx/blob/7b33a90d7441909664a920b0687db8d984ac314b/mainline/jessie/Dockerfile#L22
```
So, this PR is breaking `/dev/stdout` and `/dev/stderr` for the container.

cyphar commented:
I know what's happening, it's just really frustrating and I'm not sure what the "right way" of dealing with it is (no doubt Docker will have to deal with this crap too when they switch to consoles). Basically, the issue is that they are pointing to `/dev/std{out,err}` which is a symlink to `/proc/1/fd/{1,2}`. Since those file descriptors are going to be either sockets or `ptys`, a program that just tries to open them is going to have a bad time. I think `pty`s should work fine, but `terminal = false` containers are simply not going to work like that.

Hmm...

mrunalp commented:
@cyphar Can we use a pipe as an intermediate?

cyphar commented:
You mean for `terminal = false`? Possibly, I'd have to play with it to see if it'll work.

mrunalp commented:
@cyphar Yeah, for terminal=false we can have pipes dup'ed to the container process fds 0,1,2.

mrunalp commented:
Simple replacement of socketpair by pipe worked with manual testing for me. However, we will need to create separate pipes for stdout/stderr/stdin to be able to sift that apart in the logs. I will play with this more today with k8s to see how it goes.

mrunalp commented:
I tested my patch with nginx/httpd pods and they come up fine. Here is the patch:
```
@@ -213,9 +213,11 @@ int main(int argc, char *argv[])
                 * everything.
                 *
                 * TODO: Maybe this should be done with pipe(2)s?
-                */
-               if (socketpair(AF_LOCAL, SOCK_STREAM, 0, fds) < 0)
+                *if (socketpair(AF_LOCAL, SOCK_STREAM, 0, fds) < 0)
                        pexit("Failed to create runtime_mfd socketpair");
+                */
+               if (pipe(fds) < 0)
+                       pexit("Failed to create pipe");
 
                mfd = fds[0];
                runtime_mfd = fds[1];
```

We still need to add separate pipe's for stderr/stdin and add decoration.

mrunalp commented:
This tree is rebased and has that commit on top if anyone wants to test: https://github.com/mrunalp/cri-o/commits/pr162

mrunalp commented:
The tip of my branch now has code for RFC 3339 timestamps as well :)

mrunalp commented:
I got the logs working end to end (pending stderr separation):
```
[root@dhcp-16-123 kubernetes]# ./cluster/kubectl.sh logs httpd-3531205961-clk12
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 10.88.0.233. Set the 'ServerName' directive globally to suppress this message
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 10.88.0.233. Set the 'ServerName' directive globally to suppress this message
[Mon Apr 03 23:48:49.235381 2017] [mpm_event:notice] [pid 1:tid 139827170274120] AH00489: Apache/2.4.25 (Unix) configured -- resuming normal operations
[Mon Apr 03 23:48:49.235438 2017] [core:notice] [pid 1:tid 139827170274120] AH00094: Command line: 'httpd -D FOREGROUND'
10.88.0.1 - - [03/Apr/2017:23:49:02 +0000] "GET / HTTP/1.1" 200 45
```

@cyphar Could you prepare this PR for commit by increasing the cyclo count? We can add other fixes from my branch after that. This will unblock a lot of e2e tests and we can release 0.2 :)

cyphar commented:
Awesome. Sorry for letting this PR stagnate for so long. Things have been quite hectic over here. I should probably stop taking on so many projects. :wink:

I'll fix this up today (if it isn't done, I give you leave to harass me online to remind me to do it :stuck_out_tongue:).

cyphar commented:
Alright, I've rebased, bumped cyclo 80, taken your patches and modified them slightly (added comments for future refactorings). Everything should be ready now @mrunalp.

I'd really like it if we could move to runC `1.0.0-rc3` soon because I personally hate pinning git commits of our dependencies. However we first need to update `runtime-tools` to support generation of `v1.0.0-rc5` runtime configurations.

runcom commented:
The rebase is missing https://github.com/kubernetes-incubator/cri-o/pull/425 which is needed to properly test this PR. However, after you apply #425 everything seems to be working properly :tada: 
```
12:31:44 [src/k8s.io/kubernetes] ‹master› k run redistest --image=redis:alpine
deployment "redistest" created
12:31:52 [src/k8s.io/kubernetes] ‹master› k get pods
NAME                        READY     STATUS                            RESTARTS   AGE
nginx-3862752392-ttls5      1/1       Running                           0          1m
redis-10021985-2hc8q        0/1       rpc error: code = 11 desc = EOF   0          1m
redistest-780660705-qf1p2   1/1       Running                           0          3s
12:31:55 [src/k8s.io/kubernetes] ‹master› k logs redistest-780660705-qf1p2
1:C 04 Apr 10:31:53.443 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
1:M 04 Apr 10:31:53.445 # You requested maxclients of 10000 requiring at least 10032 max file descriptors.
1:M 04 Apr 10:31:53.445 # Server can't set maximum open files to 10032 because of OS error: Operation not permitted.
1:M 04 Apr 10:31:53.445 # Current maximum open files is 1024. maxclients has been reduced to 992 to compensate for low ulimit. If you need higher maxclients increase 'ulimit -n'.
                _._
1:M 04 Apr 10:31:53.446 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:M 04 Apr 10:31:53.446 # Server started, Redis version 3.2.8
1:M 04 Apr 10:31:53.446 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled2017-04-04T12:31:53+02:00 stdout ' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.

12:31:23 [src/k8s.io/kubernetes] ‹master› k logs nginx-3862752392-ttls5
10.88.0.1 - - [04/Apr/2017:10:31:33 +0000] "GET / HTTP/1.1" 200 612 "-" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" "-"
2017/04/04 10:31:33 [error] 7#7: *1 open() "/usr/share/nginx/html/favicon.ico" failed (2: No such file or directory), client: 10.88.0.1, server: localhost, request: "GET /favicon.ico HTTP/1.1", host: "10.88.9.97", referrer: "http://10.88.9.97/"
10.88.0.1 - - [04/Apr/2017:10:31:33 +0000] "GET /favicon.ico HTTP/1.1" 404 572 "http://10.88.9.97/" "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" "-"
```

runcom commented:
LGTM ping @mrunalp @sameo

cyphar commented:
DW, I'll rebase again.

sameo commented:
Testing now on my k8s 1.6 cluster.

cyphar commented:
As an aside, is there a reason that our tests aren't mutually independent? If one of the tests fails then all of the subsequent tests will fail because the first test that failed didn't clean up properly...

sameo commented:
For some reasons the `etcd` container is not starting properly with this PR. I'm trying to fetch ocid logs for that atm.

sameo commented:
```
Apr 04 15:00:45 k8s-master ocid[1123]: time="2017-04-04 15:00:45.466380653+02:00" level=debug msg="OCI process args [etcd --listen-client-urls=http://127.0.0.1:2379 --advertise-client-urls=http://127.0.0.1:2379 --data-dir=/var/lib/etcd]"
Apr 04 15:00:45 k8s-master ocid[1123]: time="2017-04-04 15:00:45.466396860+02:00" level=debug msg="CONTAINER USER: 0"
Apr 04 15:00:45 k8s-master ocid[1123]: time="2017-04-04 15:00:45.466549081+02:00" level=debug msg="UID: 0, GID: 0, Groups: [10]"
Apr 04 15:00:45 k8s-master ocid[1123]: time="2017-04-04 15:00:45.467023984+02:00" level=debug msg="running conmon: /usr/local/libexec/ocid/conmon" args=[-c kube-system-etcd-k8s-master-0-etcd-0 -r /usr/local/sbin/runc -b /var/run/containers/aufs-containers/c47440404af27e39d8fe4ec948a8765319442bc40a5b1c67e09656c29cd4e01c/userdata -p /var/run/containers/aufs-containers/c47440404af27e39d8fe4ec948a8765319442bc40a5b1c67e09656c29cd4e01c/userdata/pidfile -l /var/log/pods/7075157cfd4524dbe0951e00a8e3129e/etcd_0.log]
Apr 04 15:00:45 k8s-master ocid[1123]: [conmon:e]: Failed to open log file No such file or directory

```

cyphar commented:
@sameo So k8s passes relative paths for `log_path`? Can you confirm whether this directory exists: `/var/log/pods/7075157cfd4524dbe0951e00a8e3129e/`? Also, what is the sandbox's logDirectory?

sameo commented:
@cyphar this is the etcd container k8s uses, and it seems to be using a relative path.

Some more debug:
```
Apr 04 15:13:41 k8s-master ocid[1123]: time="2017-04-04 15:13:41.384029916+02:00" level=debug msg="updating sandbox and container information"
Apr 04 15:13:41 k8s-master ocid[1123]: time="2017-04-04 15:13:41.384294372+02:00" level=warning msg="requested logPath for ctr id 3a6184c6ecd27457cb4c33317be84348eec754864a26f22f0e
8a7d9bbd26606f is a relative path: etcd_0.log"
Apr 04 15:13:41 k8s-master ocid[1123]: time="2017-04-04 15:13:41.384338889+02:00" level=debug msg="setting container's log_path" ctr.logfile="etcd_0.log" log_path="/var/log/pods/70
75157cfd4524dbe0951e00a8e3129e/etcd_0.log" sbox.logdir="/var/log/pods/7075157cfd4524dbe0951e00a8e3129e"
```

```
ls -lh /var/log/pods/7075157cfd4524dbe0951e00a8e3129e/etcd_0.log 
lrwxrwxrwx 1 root root 165 Mar 16 13:50 /var/log/pods/7075157cfd4524dbe0951e00a8e3129e/etcd_0.log -> /var/lib/docker/containers/df19314697ecce0c9c30754f97ce7947d231fd4b04424b16448d1a28c3075876/df19314697ecce0c9c30754f97ce7947d231fd4b04424b16448d1a28c3075876-json.log
```

And `/var/lib/docker/containers/df19314697ecce0c9c30754f97ce7947d231fd4b04424b16448d1a28c3075876/df19314697ecce0c9c30754f97ce7947d231fd4b04424b16448d1a28c3075876-json.log` does not exist.

cyphar commented:
Okay, so the actual issue is that (for some reason) the path is a symlink? Is there any chance this is something that k8s is doing behind our backs (and potentially a bug if you're using the CRI mode) -- none of our code should be creating symlinks like that.

sameo commented:
@cyphar 
> none of our code should be creating symlinks like that.

Especially since this is a link to a docker path. I'm investigating further.

cyphar commented:
@sameo It looks like this is intentional (from `pkg/kubelet/kuberuntime/kuberuntime_container.go`):

```
	// Symlink container logs to the legacy container log location for cluster logging
	// support.
	// TODO(random-liu): Remove this after cluster logging supports CRI container log path.
	containerMeta := containerConfig.GetMetadata()
	sandboxMeta := podSandboxConfig.GetMetadata()
	legacySymlink := legacyLogSymlink(containerID, containerMeta.Name, sandboxMeta.Name,
		sandboxMeta.Namespace)
	containerLog := filepath.Join(podSandboxConfig.LogDirectory, containerConfig.LogPath)
	if err := m.osInterface.Symlink(containerLog, legacySymlink); err != nil {
		glog.Errorf("Failed to create legacy symbolic link %q to container %q log %q: %v",
			legacySymlink, containerID, containerLog, err)
	}
```

Now, we could work around this on our end by removing the file before we open it for writing. But I don't really like that idea all that much.

sameo commented:
@cyphar 
> Now, we could work around this on our end by removing the file before we open it for writing. But I don't really like that idea all that much.

I agree, but otoh this is a kubernetes bug. It certainly works with the current docker CRI shim but breaks our implementation. So I'm not against removing the file ourselves if we see it is a dead link.

cyphar commented:
I've opened kubernetes/kubernetes#44043 to track the actual bug. I'll add a work-around but will mention that we can remove it once that issue is fixed.

sameo commented:
@cyphar tested with 6045c5f on my k8s 1.6 bare metal cluster and it works fine for me now. Thanks a lot.

sameo commented:
LGTM

cyphar commented:
This rebase just fixes up the lint script a bit more. There was no significant code change (it fixes a NULL deref).

mrunalp commented:
I tested just before your latest push and it worked fine for httpd/nginx :+1:

mrunalp commented:
@sameo Can you retest with etcd?

mrunalp commented:
okay, retested with latest and looks good :+1:  :) will wait for tests to go green and @sameo retest.

sameo commented:
@mrunalp re-testing...

sameo commented:
@mrunalp @cyphar Works with etcd and an nginx deployment.

mrunalp commented:
LGTM (we can merge once tests are green)

runcom commented:
works for me as well

cyphar commented:
EUGH. I made a typo in the lint command...

runcom commented:
(Travis isn't giving a single fu*k about our tests apparently)

cyphar commented:
I think it's because Travis limits the number of jobs for each **organisation** not repo, so there's some other job being run on behalf of @kubernetes-incubator that is blocking us. FFS.

cyphar commented:
Oh goddammit. `go vet` changed their output format between Go releases. Why are `lint` tests breaking things...

cyphar commented:
:tada: The tests are green! :lizard: :green_heart:

cyphar commented:
Merging.

mrunalp commented:
@cyphar Yay :tada:

runcom commented:
:tada:

rhatdan commented:
Does this help our k8s test matrix?

mrunalp commented:
Yes. This and other patches in progress for logging helped pass a lot of the tests.


> On Apr 8, 2017, at 1:09 PM, Daniel J Walsh <notifications@github.com> wrote:
> 
> Does this help our k8s test matrix?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
